guide/adopting-the-configuration-cache.md

```diff
@@ -0,0 +1,384 @@
+|         `--- :compileJava                                    (org.gradle.api.tasks.compile.JavaCompile)
+`--- :check                                                    (org.gradle.api.DefaultTask)
+   +--- :checkstyleMain                                      (org.gradle.api.plugins.quality.Checkstyle)
+   |    +--- ...
+   +--- :checkstyleTest                                      (org.gradle.api.plugins.quality.Checkstyle)
+   |    +--- ...
+   +--- :spotlessCheck                                       (org.gradle.api.DefaultTask)
+   |    `--- ...
+   +--- :test                                                (org.gradle.api.tasks.testing.Test)
+   |    `--- ...
+   ...
+```
+
+When you first run a gradle task, the task graph is serialized and stored on disk. Upon a subsequent run of the same task, If none of these inputs have changed, Gradle skips the configuration phase entirely, loads the task graph from disk, and goes straight to the execution phase.
+
+The configuration phase typically runs faster than execution since heavy work belongs in the latter. However, without the cache, configuration phase is rerun with every single Gradle run. If you run a unit test multiple times, configuration is repeated, reproducing the same task graph every time. When the task itself is light, configuration can take up a large fraction of the latency. Configuration caching solves this by storing and reusing configuration results between runs, eliminating redundant work and speeding up iteration cycles.
+
+
+
+## Finding Configuration Cache problems
+
+To find Configuration Cache problems:
+
+1. Run `./gradlew build --configuration-cache`
+2. Check the output for any problems. If none appear, great! Your build is compatible.
+3. If not, fix the problems.
+4. Run `./gradlew build --configuration-cache` again. In our experience, this reveals additional issues not seen in the first run.
+
+The problems typically fall into three categories:
+
+1. **external process started `/usr/bin/git --version`**
+2. **cannot serialize object of type `org.gradle.api.internal.project.DefaultProject`, a subtype of `org.gradle.api.Project`, as these are not supported with the configuration cache**
+3. **invocation of `Task.project` at execution time is unsupported**
+
+The first error occurs when you start external processes during the configuration phase. Generally, it is preferred to run external processes in tasks with properly declared inputs and outputs to avoid unnecessary work when the task is UP-TO-DATE. 
+
+The second and third errors occur because Gradle tasks shouldn't take mutable types as input — `Gradle`, `Settings`, `Project`, `SourceSet`, or `Configuration`. Having these as inputs limits task concurrency (what if two tasks mutate a `Project` concurrently?) and prevents Gradle from serializing task inputs into the cache (they are too large to be serialized). 
```

Comment:
I don't think size is necessarily an issue here. It's more that you can't reason about the state and reorder and parallelise things if you're reaching into other projects and using that to configure yourself. This is the reason why it limits task concurrency (*within* a project).

It also doesn't want you mutating the big mutable state at task execution time either, you could effectively change how things are and end up with a cache that's inconsistent with reality.

Reply:
Got it, removed the inaccurate claim that size was the issue.